### ISSUE TITLE
Improvement suggestions regarding tag handling

### ISSUE THREAD

**Issue by anothertal3** [TheAnachronism/ImageDataSetTagEditor#3]
First of all: thank you for your work on this small tool which I accidentally found via reddit.

I did some testing and it appears to prove quite useful.

However, there were a couple of things which suprised me and kind of hindered the "usability flow":
- Out of habit I attempt to complete the currently selected tag with either a mouse click on the list entry or by pressing the tab-key. Neither of those worked, though. Apparently only "Enter" is supported for tag confirmation?
- It would be very useful if there was a way to directly add existing tags for a new image by simply clicking on the entry in the list to the right (or maybe on a new small button next to the entry?)
- Captions are not automatically split by "," if entered into one line. It requires a reload of the directory for that.
- When adding a new tag (e.g., via ALT+(Plus)) there's always a placeholder text "New Tag" which needs manual deletion. Maybe the placeholder could be appear highlighted already so the first keypress replaces it?

Maybe some of those suggestions might be implemented in a future release?

**Comment by TheAnachronism** [comment]
Hello there,

Thanks for using this little software 😄 

The points about the improvements, I'll have a look at what I can do. Many of them are also on a to-do list of mine, but I didn't have time to implement them yet.

**Comment by anothertal3** [comment]
Thank you for the fast response. "Time issues" is a very relatable thing :)
I'll be happy to try out newer versions when you find the time.

I've continued tagging and found a couple of further suggestions but they might be over the top for now.

For now, the biggest hindrance - at least to me - is the placeholder issue.

**Comment by Hamsterun** [comment]
Gonna leave the same comment as for the other tool "BooruDatasetTagManager":

"Would like to define Dropdowns with a custom title and add Tags, in the "all" tag section, to them for easier navigation when looking for something.
As I need specific Tags in a specific order.
So I don't have to think about the alphabetical order.
Would be neat. Thanks."

In case you find the idea useful for your tool.

**Comment by TheAnachronism** [comment]
> Would like to define Dropdowns with a custom title and add Tags, in the "all" tag section, to them for easier navigation when looking for something.

Yeah I was thinking about something like that as well. Somehow grouping or categorizing tags would allow for a far better experience of managing a dataset.
Something I'm also considering, is rethinking how the tags are actually persisted.

Something that might make sense, is to create a local database for an image set, index the actual images and store the tag definitions inside the database. Once the actual captions are needed, you can export your tagged combinations to a specific directory.
With an actual database in the background, tag management would be much easier to handle, as well as allowing for more complicated setups.
If implemented, I would go for something like a local sqlite database for each dataset you want to tag. That way you can store the database file in the root directory of the dataset and keep everything together (and even share the whole thing in a .zip file or so).

But I'm quite busy this month and will probably have a look at the entire thing in May.

**Comment by anothertal3** [comment]
All of this seems to be very reasonable and useful. However, I'd argue that input flow and usability are currently more important than a powerful tag database:
- Most user input issues won’t be affected by a new tag provider mechanism and need to be adressed anyway
- Certain usability issues are probably quite easy to change but annoying right now (especially the necessity to always manually remove the placeholder!)

But regarding the idea: grouping and hierarchies of keywords are pretty well implemented in many DAM-applications such as Lightroom. Maybe it’s possible to find some interesting ideas there? 

Let’s take, for example, the export/import of keywords. If I recall correctly, they export to a simple text file with indentations for hierarchical levels:
```
Building
	Church
		Bell Tower
	Stadium
...
```

I think that something like this might be completely sufficient for the time being. Maybe something along the lines of:
1. Program start: initialize all persistent tags from said text file and organize them in memory
2. Add new tags while tagging in memory
3. On program end: overwrite previous file with current tag hierarchy from memory (optional: rename the previous file instead for safety)

I actually don't think that you need to keep persistent references between pictures and tags in your database. This is probably rather a task for a generic image management application and I actually appreciate that you focused your tool on the actually important task: tagging for AI training via side-car file.
